Refactored psychsyn, psychsyn_critval, and mahad

camkay opened the pull request:
First off, I really appreciate your package. It has drastically improved my data screening workflow.

With respect to this pull request, I noticed that `psychsyn` could take a while to calculate the psychometric synonym scores, especially when there was a large number of items in the dataset. I identified `stats::cor(x, use = "pairwise.complete.obs")` (and specifically `use = "pairwise.complete.obs"`) as the culprit. As such, I updated the code to only use `"pairwise.complete.obs"` when the dataset is not complete. 

In terms of time savings, the screenshots included below show that, when there is missing data, the old and refactored versions of `psychsyn` perform comparably. However, when the data _is_ complete, the function provides the psychometric synonym scores in less than half the time.

**psychsyn old (complete data)**
![psychsyn_complete_old](https://user-images.githubusercontent.com/25355753/119381370-0ae6e900-bc76-11eb-9f7a-799151ff3a59.png)
**psychsyn old (incomplete data)**
![psychsyn_nas_old](https://user-images.githubusercontent.com/25355753/119381385-0b7f7f80-bc76-11eb-935e-42c9822b251d.png)
**psychsyn new (complete data)**
![psychsyn_complete_new](https://user-images.githubusercontent.com/25355753/119381391-0c181600-bc76-11eb-851d-a197f77866fc.png)
**psychsyn new (incomplete data)**
![psychsyn_nas_new](https://user-images.githubusercontent.com/25355753/119381397-0c181600-bc76-11eb-97d9-a4f431381867.png)

Best,
Cameron

camkay committed:
Refactored psychsyn and psychsyn_critval

camkay committed:
Switched mahad from using psych::outlier to using stats::mahalanobis.

camkay commented:
Since writing my previous pull requests, I also updated `mahad` to use `stats::mahalanobis` instead of `psych::outlier`. This required additional code be added to the function to create the optional Q-Q plot, but I think the time savings are worth it. Specifically, the new code runs in under half the time as the previous code (see below). 

**mahad old**
![mahad_old](https://user-images.githubusercontent.com/25355753/119412353-96756f80-bca0-11eb-9782-91046b711ee9.png)

**mahad new**
![mahad_new](https://user-images.githubusercontent.com/25355753/119412381-a2f9c800-bca0-11eb-8d46-c09fc8d8fff3.png)

franciscowilhelm commented:
Dear camkay, thanks a lot for your work!

Do you have a reproducible example or could provide the code with which you generated the fabricated_dataset (missing/non-missing)?

For your suggested changes to mahaD, we would need to check how this could work with missing data, afaik the stats::mahalanobis function does not support NAs. Potentially we could fall back to the psych version as in your changes to psychsyn, or implement a modified version of sstats::mahalanobis.

camkay committed:
Updated mahad to only use stats::mahalanobis when NAs are NOT present.

camkay commented:
Not a problem!

You are correct that `stats::mahalanobis` is not working with NAs. I had primarily been using `careless_dataset_na` to test `mahad` and had falsely assumed that the `NA`s were distributed randomly throughout the data. I have committed a new version of the function that falls back to the `{psych}` version when missing data is present. This is, of course, not ideal. I think implementing a modified version of `stats::mahalanobis` would be a great idea (but I imagine it wouldn't be a top priority).

The fabricated data was just something I threw together to test `psychsyn` with similar dimensions to a dataset I am currently working with. It can be reproduced using the following code:

```
set.seed(42)

# fabricated data without NAs
data_fabricated <- data.frame(id = 1:500)

for (i in 1:100) {
  data_fabricated[, paste0("var_", i)]       <- rnorm(n = 500, mean = 0, sd = 1)
  data_fabricated[, paste0("var_", i + 100)] <- data_fabricated[, paste0("var_", i)] + (.5 * rnorm(n = 500, mean = 0, sd = 1))
}

# fabricated data with NAs
data_fabricated_nas <- data.frame(id = 1:500)

for (i in 1:100) {
  data_fabricated_nas[, paste0("var_", i)]       <- rnorm(n = 500, mean = 0, sd = 1)
  data_fabricated_nas[, paste0("var_", i + 100)] <- data_fabricated_nas[, paste0("var_", i)] + (.5 * rnorm(n = 500, mean = 0, sd = 1))
  
  na_index  <- floor(runif(50, 1, 500))
  na_index2 <- floor(runif(50, 1, 500))
  
  data_fabricated_nas[na_index,  paste0("var_", i)]       <- NA
  data_fabricated_nas[na_index2, paste0("var_", i + 100)] <- NA
}
```

camkay committed:
Used the same code to create the plots when NAs are present and when NAs are not present.

franciscowilhelm committed:
minor change to d_sq

d_sq is initialized with the wrong length, which does not cause issues but makes the code less readable.

franciscowilhelm reviewed (approved):
Thanks again camkay, pull request looks good to me.

camkay commented:
Happy to help, @franciscowilhelm!

ryentes commented:
Hi thanks for submitting the pull request. The notifications got caught by my spam filter so I haven't seen these just yet. We're going to push some more comprehensive tests, and then run this code through them. After that I'll go ahead and merge the changes.

camkay commented:
Not a problem at all, @ryentes! Pushing some more comprehensive tests is a good call. When I was writing the pull request, I kept having to come up with ad hoc tests to make sure everything was working as expected. Having formal, prewritten tests would help immensely.
